src/libopensc/card-piv.c

```diff
@@ -506,201 +839,1973 @@ static struct sc_card_driver piv_drv = {
+		goto err;
+	}
+
+	ed_ctx = EVP_CIPHER_CTX_new();
+	if (ed_ctx == NULL) {
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
+
+	/* generate same IV used to encrypt response on card */
+	if (EVP_EncryptInit_ex(ed_ctx, (*cs->cipher_ecb)(), NULL, priv->sm_session.SKenc, zeros) != 1
+			|| EVP_CIPHER_CTX_set_padding(ed_ctx,0) != 1
+			|| EVP_EncryptUpdate(ed_ctx, IV, &outli, priv->sm_session.resp_enc_counter, 16) != 1
+			|| EVP_EncryptFinal_ex(ed_ctx, discard, &outdl) != 1
+			|| outdl != 0) {
+		sc_log(card->ctx,"SM encode failed in OpenSSL");
+		piv_log_openssl(card->ctx);
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	/* some commands do not have response data */
+	if (ee.value != NULL) {
+		p = ee.value;
+		inlen = ee.len;
+		if (inlen < 17 || *p != 0x01) { /*padding indicator is required */
+			sc_log(card->ctx, "SM padding indicator not 0x01");
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+		p++; /* skip padding indicator */
+		inlen --;
+
+		if ((inlen % 16) != 0) {
+			sc_log(card->ctx,"SM encrypted data not multiple of 16");
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+		/*
+		 * Encrypted data has 1 to 16 pad bytes, so may be 1 to 16 bytes longer
+		 * then expected. i.e. plain->resp and resplen.So will do last block
+		 * and recombine.
+		 */
+
+		inlen -= 16;
+		if (plain->resplen < (unsigned) inlen) {
+			sc_log(card->ctx, "SM response will not fit in resp,resplen");
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+		q = plain->resp;
+
+		/* first round encryptes counter with zero IV, and does not save the output */
+		if (EVP_CIPHER_CTX_reset(ed_ctx) != 1
+				|| EVP_DecryptInit_ex(ed_ctx, (*cs->cipher_cbc)(), NULL, priv->sm_session.SKenc, IV) != 1
+				|| EVP_CIPHER_CTX_set_padding(ed_ctx,0) != 1
+				|| EVP_DecryptUpdate(ed_ctx, q ,&outl, p, inlen) != 1
+				|| EVP_DecryptUpdate(ed_ctx, lastb, &outll, p + inlen, 16 ) != 1
+				|| EVP_DecryptFinal_ex(ed_ctx, discard, &outdl) != 1
+				|| outdl != 0
+				|| outll != 16) {  /* should not happen */
+			sc_log(card->ctx,"SM _decode failed in OpenSSL");
+			piv_log_openssl(card->ctx);
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+		/* unpad last block and get bytes in last block */
+		for (i = 15; i >=  0 ; i--) {
+			if (lastb[i] == 0x80)
+				break;
+			if (lastb[i] == 0x00)
+				continue;
+			sc_log(card->ctx, "SM Padding not correct");
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+		if (lastb[i] != 0x80) {
+			sc_log(card->ctx, "SM Padding not correct");
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+		/* will response fit in plain resp buffer */
+		if ((unsigned)inlen + i > plain->resplen || plain->resp == NULL) {
+			sc_log(card->ctx,"SM response bigger then resplen");
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+		/* copy bytes in last block  if any */
+		memcpy(plain->resp + inlen, lastb, i);
+		plain->resplen = inlen + i;
```

Comment:
This should be set to zero when `ee.value == NULL`.

Reply:
Lines 1404-1406 are in the if statement at line 1332  `if (ee.value != NULL) {` 
  `plain->resp` and `plain->resplen` are set by caller. The size of the response is checked but I am adding check for pointer: https://github.com/OpenSC/OpenSC/pull/2053/commits/cb81f6ccaa9b794151cd2d3a5119231d884d85d1

Do you have a PIV card or applet to test PIV  Secure Messaging?